pkg/ruler/rules/store.go

```diff
@@ -0,0 +1,106 @@
+package rules
+
+import (
+	"context"
+	"errors"
+
+	"github.com/cortexproject/cortex/pkg/configs"
+	"github.com/cortexproject/cortex/pkg/configs/client"
+
+	"github.com/prometheus/prometheus/pkg/rulefmt"
+)
+
+var (
+	// ErrGroupNotFound is returned if a rule group does not exist
+	ErrGroupNotFound = errors.New("group does not exist")
+	// ErrGroupNamespaceNotFound is returned if a namespace does not exist
+	ErrGroupNamespaceNotFound = errors.New("group namespace does not exist")
+	// ErrUserNotFound is returned if the user does not currently exist
+	ErrUserNotFound = errors.New("no rule groups found for user")
+)
+
+// RuleStore is used to store and retrieve rules
+type RuleStore interface {
+	ListAllRuleGroups(ctx context.Context) (map[string]RuleGroupList, error)
+}
+
+// RuleGroupList contains a set of rule groups
+type RuleGroupList []*RuleGroupDesc
+
+// Formatted returns the rule group list as a set of formatted rule groups mapped
+// by namespace
+func (l RuleGroupList) Formatted() map[string][]rulefmt.RuleGroup {
+	ruleMap := map[string][]rulefmt.RuleGroup{}
+	for _, g := range l {
+		if _, exists := ruleMap[g.Namespace]; !exists {
+			ruleMap[g.Namespace] = []rulefmt.RuleGroup{FromProto(g)}
+			continue
+		}
+		ruleMap[g.Namespace] = append(ruleMap[g.Namespace], FromProto(g))
+
+	}
+	return ruleMap
+}
+
+// ConfigRuleStore is a concrete implementation of RuleStore that sources rules from the config service
+type ConfigRuleStore struct {
```

Comment:
This implementation looks good to me. @bboreham are you ok with abstracting `since` into a concrete type that lives behind the interface? That way a full set of rules can be returned on each poll. However, only new rules will be polled from the config service? I don't think we can work around polling the entire ruleset currently since the horizontally sharded ruler will need to hash each rule group to ensure it is evaluating the appropriate set of rules.

Reply:
The thing that I worry about is like this: say we have 10,000 rule groups across all tenants, and one tenant changes one of them, does the program do 10,000 things or 1 thing?

Reply:
Feel free to correct any errors @jtlisi.

As it is currently designed, once a polling cycle each ruler will calculate a hash for every rule group to determine which groups it should process locally.  This will happen 10,000 times per ruler in your scenario.

Next it will this subset of the rule groups and compare them to locally stored files on disk only updating those files on disk that have changed.  This will happen 10,000 / n times per ruler.  Where n is the number of rulers.

Then, if any files have changed or been added for a given user, it will clear the old prometheus rules manager for that user and build a new one pointed at the new set of files.

Even if we do not perform this process once a polling cycle we would need to at least do this process when certain events happen (such as a ruler joining or leaving the ring).  I believe @jtlisi was preferring the straightforward nature of this approach.